stdlib/public/Platform/musl.modulemap

```diff
@@ -2,110 +2,876 @@
 //
 // This source file is part of the Swift.org open source project
 //
-// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
 // Licensed under Apache License v2.0 with Runtime Library Exception
 //
 // See https://swift.org/LICENSE.txt for license information
 // See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
 //
 //===----------------------------------------------------------------------===//
 
-module SwiftMusl [system] {
-  header "stdc-predef.h"
+module _musl [system] {
+  config_macros _GNU_SOURCE, _BSD_SOURCE, _XOPEN_SOURCE, _POSIX_SOURCE, _POSIX_C_SOURCE, _LARGEFILE64_SOURCE, _ALL_SOURCE, _DEFAULT_SOURCE, __STRICT_ANSI__
+
   header "features.h"
 
-  // C standard library
+  header "bits/musldefs.h"
+
+  exclude header "bits/alltypes.h"
+
+  // Type declarations live in bits
+  module types {
+    umbrella "bits/types"
+      module * {
+      export *
+    }
+  }
+}
+
+module stdc_assert [system] {
+  config_macros NDEBUG
+  textual header "assert.h"
+  export *
```

Comment:
textual headers don't export * because they don't build with their module and thus don't import anything to be exported. Another issue with textual headers is they need to not declare anything, otherwise the declarations will be written to multiple modules, with potentially different definitions. assert.h is particularly snarly though, it has to define `assert`, but anything that uses `assert` in the OS/SDK needs to `undef assert` so that its module doesn't bake in a declaration for `assert` that's potentially incompatible with other modules.

Reply:
There's a limit to the amount of change I can (easily) make to the musl headers, so to the extent that this problem exists it's going to have to remain a problem for now; even if we can upstream changes to resolve this, experience so far suggests it's quite difficult to get and sustain the attention of musl's (sole) maintainer, and he's the only one who actually merges changes as far as I can tell.

Reply:
(The changes I've made, for instance to put all the types in separate headers in `bits`, are done automatically by a script.)

Reply:
Yeah, sure, but we can still drop the `export *` since it doesn't do anything. 😄